### PR TITLE
[SPARK-23751][ML][PySpark] Kolmogorov-Smirnoff test Python API in pyspark.ml

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Kolmogorov-Smirnoff test Python API in `pyspark.ml`

**Note**  API with `CDF` is a little difficult to support in python. We can add it in following PR.

## How was this patch tested?

doctest

### CODE REVIEW

File: python/pyspark/ml/stat.py [hunk 1]
```diff
@@ -134,6 +134,63 @@ def corr(dataset, column, method="pearson"):
...
+    @since("2.4.0")
+    def test(dataset, sampleCol, distName, *params):
+        """
+        Perform a Kolmogorov-Smirnov test using dataset.
+        """
+        sc = SparkContext._active_spark_context
+        javaTestObj = _jvm().org.apache.spark.ml.stat.KolmogorovSmirnovTest
+        dataset = _py2java(sc, dataset)
+        params = [float(param) for param in params]
+        return _java2py(sc, javaTestObj.test(dataset, sampleCol, distName,
```
[the remark this answers is not in the document]
Comment (yogeshg) [reply python/pyspark/ml/stat.py]: Looks like we should be able to use `pyspark.ml.common.callJavaFunc`
```python
from pyspark.ml.common import _java2py, _py2java, callJavaFunc
...

return callJavaFunc(sc, javaTestObj, dataset, sampleCol, distName, *params)
```

but I am not fully sure if `_jvm().PythonUtils.toSeq()` method will still be required. 🤞

**Review by yogeshg — CHANGES_REQUESTED** [review]
Looks great, thanks for the work! I have small suggestions with documentation and jvm calls. Otherwise looks good to me :)

File: python/pyspark/ml/stat.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yogeshg) [reply python/pyspark/ml/stat.py]: thanks for checking this out! current usage sounds fair!

**Comment by yogeshg** [comment]
lgtm, I'll defer to @jkbradley

File: mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala [hunk 2]
```diff
@@ -59,7 +59,7 @@ object KolmogorovSmirnovTest {
    * distribution of the sample data and the theoretical distribution we can provide a test for the
    * the null hypothesis that the sample data comes from that theoretical distribution.
    *
-   * @param dataset a `DataFrame` containing the sample of data to test
+   * @param dataset A dataset or a dataframe containing the sample of data to test
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: nit: It's nicer to keep single back quotes ``` `DataFrame` ``` to make these show up as code in docs for clarity.  No need to get rid of that.

File: mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala [hunk 3]
```diff
@@ -102,10 +102,11 @@ object KolmogorovSmirnovTest {
    */
   @Since("2.4.0")
   @varargs
-  def test(dataset: DataFrame, sampleCol: String, distName: String, params: Double*): DataFrame = {
+  def test(dataset: Dataset[_], sampleCol: String, distName: String, params: Double*)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: nit: This doesn't fit scala style; please get familiar with the style we use for multi-line function headers!  Just check out other parts of MLlib for examples.

File: python/pyspark/ml/stat.py [hunk 4]
```diff
@@ -134,6 +134,65 @@ def corr(dataset, column, method="pearson"):
...
+
+    Conduct the two-sided Kolmogorov Smirnov (KS) test for data sampled from a continuous
+    distribution.
+
+    By comparing the largest difference between the empirical cumulative
+    distribution of the sample data and the theoretical distribution we can provide a test for the
+    the null hypothesis that the sample data comes from that theoretical distribution.
+
+    :param dataset:
+      a dataset or a dataframe containing the sample of data to test.
```
Comment (jkbradley) [inline python/pyspark/ml/stat.py]: nit: dataset -> Dataset, dataframe -> DataFrame (It's nice to write class names the way they are defined.)

File: python/pyspark/ml/stat.py [hunk 5]
```diff
@@ -134,6 +134,65 @@ def corr(dataset, column, method="pearson"):
...
+    :return:
+      A dataframe that contains the Kolmogorov-Smirnov test result for the input sampled data.
+      This DataFrame will contain a single Row with the following fields:
+      - `pValue: Double`
+      - `statistic: Double`
+
+    >>> from pyspark.ml.stat import KolmogorovSmirnovTest
+    >>> dataset = [[-1.0], [0.0], [1.0]]
+    >>> dataset = spark.createDataFrame(dataset, ['sample'])
+    >>> ksResult = KolmogorovSmirnovTest.test(dataset, 'sample', 'norm', 0.0, 1.0).collect()[0]
```
Comment (jkbradley) [inline python/pyspark/ml/stat.py]: nit: use first() instead of collect()[0]

File: python/pyspark/ml/stat.py [hunk 6]
```diff
@@ -134,6 +134,65 @@ def corr(dataset, column, method="pearson"):
...
+    .. note:: Experimental
+
+    Conduct the two-sided Kolmogorov Smirnov (KS) test for data sampled from a continuous
+    distribution.
+
+    By comparing the largest difference between the empirical cumulative
+    distribution of the sample data and the theoretical distribution we can provide a test for the
+    the null hypothesis that the sample data comes from that theoretical distribution.
+
+    :param dataset:
```
Comment (jkbradley) [inline python/pyspark/ml/stat.py]: I see you're following the example of ChiSquareTest, but this Param documentation belongs with the test method, not the class.  Could you please shift it?  (Feel free to correct it for ChiSquareTest here or in another PR.)

File: mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala [hunk 7]
```diff
@@ -81,7 +81,7 @@ object KolmogorovSmirnovTest {
    * Java-friendly version of `test(dataset: DataFrame, sampleCol: String, cdf: Double => Double)`
    */
   @Since("2.4.0")
-  def test(dataset: DataFrame, sampleCol: String,
+  def test(dataset: Dataset[_], sampleCol: String,
     cdf: Function[java.lang.Double, java.lang.Double]): DataFrame = {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: I guess I missed this before.  Would you mind fixing the scala style here too?

File: python/pyspark/ml/stat.py [hunk 8]
```diff
@@ -134,6 +134,65 @@ def corr(dataset, column, method="pearson"):
...
+      a dataset or a dataframe containing the sample of data to test.
+    :param sampleCol:
+      Name of sample column in dataset, of any numerical type.
+    :param distName:
+      a `string` name for a theoretical distribution, currently only support "norm".
+    :param params:
+      a list of `Double` values specifying the parameters to be used for the theoretical
```
Comment (jkbradley) [inline python/pyspark/ml/stat.py]: I realized we should list what the parameters are, both here and in the Scala docs.

**Review by jkbradley — COMMENTED** [review]
Thanks for the PR!  My comments are either style nits or doc improvements.

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

File: python/pyspark/ml/stat.py [hunk 9]
```diff
@@ -127,13 +113,86 @@ class Correlation(object):
...
+    0.175
+
+    .. versionadded:: 2.4.0
+
+    """
+    @staticmethod
+    @since("2.4.0")
+    def test(dataset, sampleCol, distName, *params):
+        """
+        Perform a Kolmogorov-Smirnov test using dataset.
```
Comment (jkbradley) [inline python/pyspark/ml/stat.py]: Can you please make this match the text in the Scala doc?

**Review by jkbradley — COMMENTED** [review]
My only remaining comments are about being more careful in matching the docs between Scala and Python.

**Comment by jkbradley** [comment]
LGTM
Thanks for the PR!
Merging with master

File: mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala [hunk 10]
```diff
@@ -81,32 +81,37 @@ object KolmogorovSmirnovTest {
...
   @Since("2.4.0")
-  def test(dataset: DataFrame, sampleCol: String,
-    cdf: Function[java.lang.Double, java.lang.Double]): DataFrame = {
-    val f: Double => Double = x => cdf.call(x)
-    test(dataset, sampleCol, f)
+  def test(
+      dataset: Dataset[_],
+      sampleCol: String,
+      cdf: Function[java.lang.Double, java.lang.Double]): DataFrame = {
+    test(dataset, sampleCol, (x: Double) => cdf.call(x))
```
Comment (ueshin) [inline mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: Seems like this broke Scala 2.12 build again. Can you fix it or shall I? Thanks!

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: ah.... it's my fault. Sorry!

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: [SPARK-23751][FOLLOW-UP] fix build for scala-2.12

Comment (jkbradley) [reply mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: This is the 2nd time I've merged something which broke the Scala 2.12 build.  If we care about Scala 2.12, do you know if there are any efforts to add PR builder tests for 2.12?

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: I can build locally against scala-2.12 to check it. :)

**Pull request closed without merging**